### PR TITLE
HyperbandPruner

### PR DESCRIPTION
**Pull request by c-bata**
https://arxiv.org/abs/1603.06560

Hyperband requires two inputs:

* R: the maximum amount of resource that can be allocated to a single configuration
* η: an input that controls the proportion of configurations discarded in each round of SuccessiveHalving

Functions:

* `run_then_return_val_loss(t, r)` – a function that takes a hyperparameter configuration t and resource allocation r as input and returns the validation loss after training the configuration for the allocated resources.
* `top_k(configs, losses, k)` – a function that takes a set of configurations as well as their associated losses and returns the top k performing configurations.

### CODE REVIEW

**Comment by c-bata** [comment]
refs https://github.com/optuna/optuna/pull/785

**Comment by c-bata** [comment]
HyperbandPruner implementation of Optuna use a bit tricky technique like:

https://github.com/optuna/optuna/pull/809/files#diff-45eb5bd2359f1b7f13b35b43caecb92bR119-R153

Go doesn't have this kind of flexibility. So I think following approach is better.

1. Add `TrialFilter` interface. (it seems to need to consider more clear naming)
1. `hyperband.Pruner` implements a `TrialFilter` interface.
1. Check `filter, ok := pruner.(TrialFilter)`

**Comment by c-bata** [comment]
It's difficult to implement Hyperband in the current design of Goptuna.
